modules/core/04-channel/keeper/upgrade.go

```diff
@@ -699,47 +699,50 @@ func (k Keeper) startFlushUpgradeHandshake(
 		)
 	}
 
+	if err := k.checkForUpgradeCompatibility(ctx, proposedUpgradeFields, counterpartyUpgrade); err != nil {
+		return types.NewUpgradeError(channel.UpgradeSequence, err)
+	}
+
+	return nil
+}
+
+// checkForUpgradeCompatibility checks performs stateful validation of self upgrade fields relative to counterparty upgrade.
+func (k Keeper) checkForUpgradeCompatibility(ctx sdk.Context, proposedUpgradeFields types.UpgradeFields, counterpartyUpgrade types.Upgrade) error {
```

Comment:
is this stateful validation? we pass in both `proposedUpgrade` and `counterpartyUpgrade`. not really a blocker just a question 🤔

Reply:
it's stateful in that we fetch the connection state and check if it's open (something we can't do in validate basic)